**Importing all the necessary Libraries**

In [4]:
import pandas as pd
import numpy as np

**Create a dataframe of the dataset**

In [5]:
df=pd.read_csv(r"C:\Users\omen\Documents\PythonCodes\Sem 6\Tests\Creditcard_data.csv")
X=df.drop(columns=['Class'])
Y=df['Class']
print(X.shape)
print(Y.shape)

(772, 30)
(772,)


**Checking Whether there are any null values or not**

In [6]:
df.isnull().values.any()

False

**Checking the balance of the dataset**

In [7]:
n1 = df[df['Class']==1].count().max()
n0 = df[df['Class']==0].count().max()
print(n1," objects belong to class 1")
print(n0," objects belong to class 0")

9  objects belong to class 1
763  objects belong to class 0


*Here, the dataset is imbalanced because there are only 9 data objects belonging to class 1 while the rest 763 belong to class 0*

**Splitting the dataset into two different sets of the individual classes**

In [8]:
fraud = df[df['Class']==1]
normal = df[df['Class']==0]

*To fix this imbalance, we use the techniques of undersampling and oversampling to create a balanced dataset*

In [9]:
# Undersampling

from imblearn.under_sampling import NearMiss

nm = NearMiss()
X1_res,Y1_res = nm.fit_resample(X,Y)

In [10]:
# Oversampling

from imblearn.over_sampling import RandomOverSampler

ROS = RandomOverSampler()
X2_res,Y2_res = ROS.fit_resample(X,Y)

In [11]:
# Error = 0.05 for simple random sampling
n = (1.96*1.96*0.5*0.5)//(0.05**2)

In [12]:
df2=X2_res
df2['Class']=Y2_res
df2

Time        V1        V2        V3        V4        V5        V6  \
0        0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1        0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2        1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3        1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4        2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...    ...       ...       ...       ...       ...       ...       ...   
1521   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1522   539 -1.738582  0.052740  1.187057 -0.656652  0.920623 -0.291788   
1523   484 -0.928088  0.398194  1.741131  0.182673  0.966387 -0.901004   
1524   574  1.257719  0.364739  0.306923  0.690638 -0.357792 -1.067481   
1525     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   

            V7        V8        V9  ...       V21       V22       V23  \
0     0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474   
1    -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
2     0.791461  0.247676 -1.514654  ...  0.247998  0.771679  0.909412   
3     0.237609  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321   
4     0.592941 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458   
...        ...       ...       ...  ...       ...       ...       ...   
1521  0.086781 -0.202836  0.035154  ... -0.287592 -0.832682  0.128083   
1522  0.269083  0.140631  0.023464  ... -0.179545 -0.192036 -0.261879   
1523  0.879016 -0.156590 -0.142117  ...  0.066353  0.281378 -0.257966   
1524  0.094272 -0.210300  0.014455  ... -0.286856 -0.820658  0.127663   
1525 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   

           V24       V25       V26       V27       V28  Amount  Class  
0     0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1    -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
2    -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3    -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4     0.141267 -0.206010  0.502292  0.219422  0.215153   69.99      0  
...        ...       ...       ...       ...       ...     ...    ...  
1521  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69      1  
1522 -0.237477 -0.335040  0.240323 -0.345129 -0.383563    1.00      1  
1523  0.385384  0.391117 -0.453853 -0.104448 -0.125765    1.00      1  
1524  0.343128  0.221120  0.094391 -0.022189  0.030944    1.29      1  
1525 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  

[1526 rows x 31 columns]

**Simple Random Sampling**

In [13]:
s1 = df2.sample(n=int(n), random_state=42)
s1

Time        V1        V2        V3        V4        V5        V6  \
1439   539 -1.738582  0.052740  1.187057 -0.656652  0.920623 -0.291788   
76      49 -0.549626  0.418949  1.729833  0.203065 -0.187012  0.253878   
1010   164  0.073497  0.551033  0.451890  0.114964  0.822947  0.251480   
660    499  1.255439  0.307729  0.292700  0.699873 -0.428876 -1.088456   
1132   574  1.257719  0.364739  0.306923  0.690638 -0.357792 -1.067481   
...    ...       ...       ...       ...       ...       ...       ...   
1481     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
756    564 -0.203837  0.532747 -0.339857 -0.730934  2.728163  3.535882   
1074   484 -0.928088  0.398194  1.741131  0.182673  0.966387 -0.901004   
867    472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
485    356  1.021392  0.155451 -0.312291  0.961407  0.548611  0.121816   

            V7        V8        V9  ...       V21       V22       V23  \
1439  0.269083  0.140631  0.023464  ... -0.179545 -0.192036 -0.261879   
76    0.500894  0.251256 -0.227985  ...  0.115062  0.418529 -0.065133   
1010  0.296319  0.139497 -0.123050  ... -0.128758 -0.381932  0.151012   
660   0.043840 -0.167739  0.128854  ... -0.294795 -0.882126  0.136846   
1132  0.094272 -0.210300  0.014455  ... -0.286856 -0.820658  0.127663   
...        ...       ...       ...  ...       ...       ...       ...   
1481 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
756   0.263680  0.919169 -0.194501  ... -0.082087 -0.271636 -0.157778   
1074  0.879016 -0.156590 -0.142117  ...  0.066353  0.281378 -0.257966   
867   0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
485   0.525318 -0.053891 -0.686867  ... -0.224762 -0.850712 -0.077845   

           V24       V25       V26       V27       V28  Amount  Class  
1439 -0.237477 -0.335040  0.240323 -0.345129 -0.383563    1.00      1  
76    0.264981  0.003958  0.395969  0.027182  0.043506   59.99      0  
1010 -1.363967 -1.389079  0.075412  0.231750  0.230171    0.99      1  
660   0.327949  0.194459  0.096516 -0.027271  0.029491    1.98      0  
1132  0.343128  0.221120  0.094391 -0.022189  0.030944    1.29      1  
...        ...       ...       ...       ...       ...     ...    ...  
1481 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
756   0.989458  0.228821 -0.545156  0.058120  0.035573   12.90      0  
1074  0.385384  0.391117 -0.453853 -0.104448 -0.125765    1.00      1  
867  -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
485  -0.679775  0.484478 -0.605930 -0.005454  0.019392  105.87      0  

[384 rows x 31 columns]

**Bootstrap Sampling**

In [14]:
# Oversampling
s2 = df2.sample(n=int(n), replace=True, random_state=42)

s2


Time        V1        V2        V3        V4        V5        V6  \
1126   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1459   529 -2.000567 -2.495484  2.467149  1.140053  2.462010  0.594262   
860      0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
1294   539 -1.738582  0.052740  1.187057 -0.656652  0.920623 -0.291788   
1130   406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
...    ...       ...       ...       ...       ...       ...       ...   
565    423  1.174846 -0.097105 -0.361083  0.914490  0.594571  1.106664   
569    426 -0.424126  0.943262  1.133354 -0.166338  0.387243 -0.030382   
1346   472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
685    517 -0.639474 -0.048355  2.452755  0.310804 -0.430963 -0.290032   
625    476 -0.867554 -0.418633  2.294927  0.219870 -0.085820 -0.839204   

            V7        V8        V9  ...       V21       V22       V23  \
1126  0.086781 -0.202836  0.035154  ... -0.287592 -0.832682  0.128083   
1459 -2.110183  0.788347  0.958809  ...  0.422452  1.195394  0.297836   
860  -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
1294  0.269083  0.140631  0.023464  ... -0.179545 -0.192036 -0.261879   
1130 -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
...        ...       ...       ...  ...       ...       ...       ...   
565  -0.114835  0.307650  0.347846  ... -0.157769 -0.388798 -0.303408   
569   0.466045  0.275041 -0.527932  ... -0.211158 -0.582126 -0.021227   
1346  0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
685   0.166889  0.006196  0.651675  ...  0.004189  0.110847  0.057008   
625  -0.724947  0.221812  0.423738  ...  0.231306  0.547470 -0.001608   

           V24       V25       V26       V27       V28  Amount  Class  
1126  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69      1  
1459 -0.857105 -0.219322  0.861019 -0.124622 -0.171060    1.50      1  
860  -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
1294 -0.237477 -0.335040  0.240323 -0.345129 -0.383563    1.00      1  
1130  0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
...        ...       ...       ...       ...       ...     ...    ...  
565  -1.737288  0.805329 -0.221755  0.018558 -0.008570   43.34      0  
569  -0.379104 -0.241854  0.106796  0.251811  0.080462    4.95      0  
1346 -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
685   0.389171 -0.449642  0.218186 -0.067664 -0.073760   59.90      0  
625   0.461131 -0.504834  1.146692  0.061651  0.146244    1.00      0  

[384 rows x 31 columns]

**Stratified Sampling**

In [15]:
z=1.96
p=0.5
E=0.05
S=0.7
sample_size=round((z**2*p*(1-p))/((E/S)**2))
s3=df2.groupby('Class', group_keys=False).apply(lambda x: x.sample(sample_size))
s3['Class'].value_counts()
s3

Time        V1        V2        V3        V4        V5        V6  \
60      41  0.986063 -0.202965 -0.492768  0.407691  0.305660 -0.230529   
596    448 -0.242497  0.891170  1.164787  1.025392  0.729123  0.663501   
362    266 -2.564961  2.470985  2.649417 -1.564256  1.794297 -0.614742   
615    463 -0.634597  0.866354  1.123836 -0.304041  1.173910  0.257118   
500    368 -0.409900  1.183088  1.598967  0.353088  0.309710 -0.312400   
...    ...       ...       ...       ...       ...       ...       ...   
1126   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1072   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1507     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
1362   529 -2.000567 -2.495484  2.467149  1.140053  2.462010  0.594262   
1136   539 -1.738582  0.052740  1.187057 -0.656652  0.920623 -0.291788   

            V7        V8        V9  ...       V21       V22       V23  \
60    0.585028 -0.208225 -0.247503  ... -0.305874 -1.216555 -0.077602   
596   0.760560 -0.004531  0.037370  ... -0.161343  0.047255 -0.302695   
362   4.185906 -3.855359  5.436633  ... -1.672706 -0.463149 -0.532466   
615   0.754325  0.171735 -0.825981  ...  0.160463  0.422873 -0.353721   
500   0.707197 -0.043206 -0.892869  ... -0.163371 -0.396155 -0.069498   
...        ...       ...       ...  ...       ...       ...       ...   
1126  0.086781 -0.202836  0.035154  ... -0.287592 -0.832682  0.128083   
1072  0.086781 -0.202836  0.035154  ... -0.287592 -0.832682  0.128083   
1507 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288   
1362 -2.110183  0.788347  0.958809  ...  0.422452  1.195394  0.297836   
1136  0.269083  0.140631  0.023464  ... -0.179545 -0.192036 -0.261879   

           V24       V25       V26       V27       V28  Amount  Class  
60   -0.741341  0.286881  0.200347 -0.075203  0.027271  169.05      0  
596  -0.931315 -0.008611 -0.187111  0.259246 -0.030635   11.99      0  
362   0.306494  0.226844 -0.365416 -0.936735 -2.733887   10.35      0  
615  -0.847192  0.206896 -0.436112  0.110134  0.102261    1.00      0  
500   0.069735 -0.298407  0.199188  0.099692  0.118617    1.98      0  
...        ...       ...       ...       ...       ...     ...    ...  
1126  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69      1  
1072  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69      1  
1507 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69      1  
1362 -0.857105 -0.219322  0.861019 -0.124622 -0.171060    1.50      1  
1136 -0.237477 -0.335040  0.240323 -0.345129 -0.383563    1.00      1  

[376 rows x 31 columns]

**Systematic Sampling**

In [16]:
n=len(df)
k=int(n**0.5)
s4=df2.iloc[::k]
s4

Time        V1        V2        V3        V4        V5        V6  \
0        0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
27      23  1.322707 -0.174041  0.434555  0.576038 -0.836758 -0.831083   
54      37  1.295668  0.341483  0.081505  0.566746 -0.110459 -0.766325   
81      52  1.147369  0.059035  0.263632  1.211023 -0.044096  0.301067   
108     73  1.162281  1.248178 -1.581317  1.475024  1.138357 -1.020373   
135     84  1.119272 -0.669639  0.803807 -0.651693 -1.395666 -0.800698   
162    103 -0.940893  1.074155  1.759398 -0.601446  0.101693 -0.188520   
189    124 -1.710935 -1.366799  2.217311  0.404714 -0.114375 -0.075942   
216    142  1.288256  0.085828 -1.179482  0.064357  2.195225  3.383363   
243    164 -0.433211  1.020835  2.019730  3.003261  0.031308  0.187063   
270    190 -0.549414  0.676861  2.151950  1.014523 -0.620012  0.076154   
297    211 -0.247827 -0.282682  1.653354 -1.014865 -0.680433  0.886364   
324    237  1.260248 -0.020172 -1.164387  0.266251  2.184886  3.607421   
351    259 -1.569485 -1.932133  1.249203 -4.434211  1.244282  0.402688   
378    275 -0.363519  0.055464  1.857571 -1.085421 -0.981918 -0.473025   
405    292  1.252189 -0.126779  0.280285  0.579416 -0.374125 -0.215217   
432    312 -0.719402 -0.124184  1.979309  0.503415 -0.732324  0.394883   
459    336 -0.895224  0.562106  2.817524 -0.718734  0.223222  0.796156   
486    356  1.586093 -1.169091 -1.350477 -2.504580  1.106389  3.135282   
513    379 -1.896099 -1.829046  1.181222  3.897846 -1.686469  1.976371   
540    406 -0.814054  1.538222  1.115690 -0.051667  0.092334 -1.013398   
567    425 -0.367058  1.035851  1.108064 -0.065459  0.366306 -0.322658   
594    446 -1.146103  1.350274  0.907209 -0.040682 -0.242920 -1.099859   
621    472 -1.100920  1.029588  1.348333 -1.362082 -0.343465 -0.671659   
648    491 -0.946412  0.609500  1.201710  0.113074 -0.210132 -0.954009   
675    511 -0.259961  0.998646  1.437975  0.038031  0.125405 -0.957775   
702    530  1.217261 -0.080646 -0.059293 -0.868862 -0.236628 -0.700159   
729    550 -1.274193  1.722263  0.429337  0.105932 -0.732006 -1.107861   
756    564 -0.203837  0.532747 -0.339857 -0.730934  2.728163  3.535882   
783    484 -0.928088  0.398194  1.741131  0.182673  0.966387 -0.901004   
810    118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
837    484 -0.928088  0.398194  1.741131  0.182673  0.966387 -0.901004   
864    118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
891    406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
918    118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
945    484 -0.928088  0.398194  1.741131  0.182673  0.966387 -0.901004   
972    529 -2.000567 -2.495484  2.467149  1.140053  2.462010  0.594262   
999    406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
1026   472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
1053   164  0.073497  0.551033  0.451890  0.114964  0.822947  0.251480   
1080   529 -2.000567 -2.495484  2.467149  1.140053  2.462010  0.594262   
1107   539 -1.738582  0.052740  1.187057 -0.656652  0.920623 -0.291788   
1134   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1161   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1188   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1215     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
1242   472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
1269   406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
1296   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
1323   529 -2.000567 -2.495484  2.467149  1.140053  2.462010  0.594262   
1350   406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
1377   406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
1404   406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
1431   164

**Cluster Sampling**

In [17]:
z=1.96
p=0.5
E=0.05
C=1.7
sample_size=round((z**2*p*(1-p))/((E/C)**2))
clusters=2
df_new=df2
N = len(df2)
K = int(N/sample_size)
data = None
for k in range(K):
    sample_k = df_new.sample(sample_size)
    sample_k["cluster"] = np.repeat(k,len(sample_k))
    df_new = df_new.drop(index = sample_k.index)
    data = pd.concat([data,sample_k],axis = 0)

random_chosen_clusters = np.random.randint(0,K,size = clusters)
s5 = data[data.cluster.isin(random_chosen_clusters)]
s5.drop(['cluster'], axis=1, inplace=True)
s5

Time        V1        V2        V3        V4        V5        V6  \
1463   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
753    562 -0.312911  0.815136  1.405270  0.070527  0.058558 -1.030855   
597    449 -0.856525  0.583290  1.389014 -0.344699  0.267594 -0.951375   
407    294 -1.097477  1.246236  0.464855 -0.178196  0.249499 -0.725058   
1161   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
...    ...       ...       ...       ...       ...       ...       ...   
1320   472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
107     71  1.331897 -0.579962  0.422606 -0.897752 -0.746254 -0.056273   
674    510  1.163271  0.141760  0.124579  0.958551 -0.159554 -0.461529   
735    552  1.238533 -0.011075  0.332073  0.086029 -0.449032 -0.653983   
1497   472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   

            V7        V8        V9  ...       V21       V22       V23  \
1463  0.086781 -0.202836  0.035154  ... -0.287592 -0.832682  0.128083   
753   0.760191 -0.182238  0.037325  ... -0.326542 -0.821322 -0.000188   
597   0.523117 -0.049229 -0.076944  ... -0.056911 -0.309940  0.167010   
407   1.197760 -0.358585 -0.153112  ... -0.050148  0.114628 -0.188974   
1161  0.086781 -0.202836  0.035154  ... -0.287592 -0.832682  0.128083   
...        ...       ...       ...  ...       ...       ...       ...   
1320  0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
107  -0.750317  0.128484 -0.964682  ... -0.039070 -0.255174  0.109333   
674   0.090759 -0.023257 -0.125187  ...  0.066320  0.089322 -0.169921   
735  -0.114230 -0.056522  0.304679  ... -0.078097 -0.206754  0.069282   
1497  0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   

           V24       V25       V26       V27       V28  Amount  Class  
1463  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69      1  
753   0.276734 -0.243046  0.049775  0.064619 -0.126739   10.99      0  
597   0.359246 -0.969651 -0.185793 -0.136897  0.233672   10.20      0  
407  -0.066727 -0.160964  0.275263  0.117647  0.272015   73.00      0  
1161  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69      1  
...        ...       ...       ...       ...       ...     ...    ...  
1320 -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
107  -0.328448  0.162254 -0.447276  0.020071  0.006231   14.48      0  
674  -0.012282  0.660756 -0.325616 -0.002210  0.010419   35.93      0  
735   0.117772  0.196515  0.760195 -0.054258  0.003194    5.95      0  
1497 -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  

[1110 rows x 31 columns]

In [18]:
samples=[]
samples.append(s1)
samples.append(s2)
samples.append(s3)
samples.append(s4)
samples.append(s5)

In [19]:
sol=pd.DataFrame(columns=['Simple-Random','Bootstrap','Stratified','Systematic','Cluster'], index=['Logistic Regression','Decision Tree','Random Forest','Naive Bayes','SVM'])

**Logistic Regression**

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [21]:
from sklearn.linear_model import LogisticRegression

for i in range(5):
    x_s=samples[i].drop('Class',axis=1)
    y_s=samples[i]['Class']
    xtrain, xtest, y_train, y_test = train_test_split(x_s ,y_s , random_state=104,test_size=0.25, shuffle=True)

    model = LogisticRegression(random_state = 0,max_iter=2000)
    model.fit(xtrain, y_train)
    y_pred = model.predict(xtest)
    acc = accuracy_score(y_test, y_pred)
    sol.iloc[0,i]=acc*100


**Decision Trees**

In [22]:
from sklearn.tree import DecisionTreeClassifier

for i in range(5):
    x_s=samples[i].drop('Class',axis=1)
    y_s=samples[i]['Class']
    xtrain, xtest, y_train, y_test = train_test_split(x_s ,y_s , random_state=104,test_size=0.25, shuffle=True)
    
    model = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
    model.fit(xtrain, y_train)
    y_pred = model.predict(xtest)
    acc = accuracy_score(y_test, y_pred)
    sol.iloc[1,i]=acc*100

**Random Forest**

In [23]:
from sklearn.ensemble import RandomForestClassifier

for i in range(5):
    x_s=samples[i].drop('Class',axis=1)
    y_s=samples[i]['Class']
    xtrain, xtest, y_train, y_test = train_test_split(x_s ,y_s , random_state=104,test_size=0.25, shuffle=True)

    model = RandomForestClassifier(n_estimators = 100)
    model.fit(xtrain, y_train)
    y_pred = model.predict(xtest)
    acc = accuracy_score(y_test, y_pred)
    sol.iloc[2,i]=acc*100

**Naive Bayes**

In [24]:
from sklearn.naive_bayes import GaussianNB

for i in range(5):
    x_s=samples[i].drop('Class',axis=1)
    y_s=samples[i]['Class']
    xtrain, xtest, y_train, y_test = train_test_split(x_s ,y_s , random_state=104,test_size=0.25, shuffle=True)

    model = GaussianNB()
    model.fit(xtrain,y_train)
    y_pred= model.predict(xtest)
    acc = accuracy_score(y_test, y_pred)
    sol.iloc[3,i]=acc*100

**Support Vector Machines**

In [25]:
from sklearn.svm import SVC

for i in range(5):
    x_s=samples[i].drop('Class',axis=1)
    y_s=samples[i]['Class']
    xtrain, xtest, y_train, y_test = train_test_split(x_s ,y_s , random_state=104,test_size=0.25, shuffle=True)

    model = SVC(kernel='rbf')
    model.fit(xtrain, y_train)
    y_pred=model.predict(xtest)
    acc = accuracy_score(y_test, y_pred)
    sol.iloc[4,i]=acc*100

In [26]:
print(sol)

                    Simple-Random  Bootstrap Stratified Systematic    Cluster
Logistic Regression     86.458333  92.708333  93.617021  93.333333  92.086331
Decision Tree           94.791667  97.916667   97.87234  73.333333  98.920863
Random Forest               100.0      100.0   98.93617  86.666667      100.0
Naive Bayes             70.833333  72.916667  84.042553  46.666667  76.978417
SVM                        65.625     78.125  72.340426  53.333333  71.582734


In [27]:
def highlight_max(s):
    is_max = s == s.max()
    return ['background-color: blue' if v else '' for v in is_max]
sol.style.apply(highlight_max, axis=1)

*Random Forest has the best overall performance*

**Shapiro-Wilk Test**

*This test is used to check the goodness of a sample*

In [28]:
from scipy.stats import shapiro
for i in range(5):
  print(shapiro(samples[i]))

ShapiroResult(statistic=0.17280685901641846, pvalue=0.0)
ShapiroResult(statistic=0.17152827978134155, pvalue=0.0)
ShapiroResult(statistic=0.16854548454284668, pvalue=0.0)
ShapiroResult(statistic=0.17053323984146118, pvalue=0.0)
ShapiroResult(statistic=0.16387814283370972, pvalue=0.0)


c:\Users\omen\AppData\Local\Programs\Python\Python310\lib\site-packages\scipy\stats\_morestats.py:1882: UserWarning: p-value may not be accurate for N > 5000.
  warnings.warn("p-value may not be accurate for N > 5000.")
